### ISSUE TITLE
Traefik fails to keep up (?) with websocket load and is killed by watchdog

### ISSUE THREAD

**Issue by Rush** [containous/traefik#4419]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?
I tried to migrate my [http-master](https://github.com/virtkick/http-master) (node-http-proxy under the hood) setup to traefik. What may be relevant is that the deployment is serving around 2100 concurrent web socket connections. with around 25% total CPU spent on proxying (+ SSL). 

### What did you expect to see?
Traefik handling the load with no sweat and no crashing (due to watchdog killing it).

### What did you see instead?
After I switched some of the traffic to Traefik it started to get real slow, and eventually it went into a watchdog crash loop. I have captured the first crash here:

https://rush.sh/gYonYrrL98CzunazlTNrMXxyEJc/traefik-crash.log
After some time I captured a graph from the dashboard:
![](http://puu.sh/CAEjw/09b6c7873d.png)

After investigating the log I have enabled the ping endpoint to see if systemd wasn't mistaken. My manual watchdog with curl indicates that indeed the process got super slow to respond to anything.
```
dragon:/home/rush # time curl http://xxx.xxx.xxx.xxx/ping
real    0m6.249s
user    0m0.026s
sys     0m0.001s
dragon:/home/rush # time curl http://xxx.xxx.xxx.xxx/ping
OK
real    0m1.417s
user    0m0.022s
sys     0m0.001s
dragon:/home/rush # time curl http://xxx.xxx.xxx.xxx/ping
OK
real    0m4.008s
user    0m0.027s
sys     0m0.000s
dragon:/home/rush # time curl http://xxx.xxx.xxx.xxx/ping
OK
real    0m10.302s
user    0m0.014s
sys     0m0.006s
```
Some of the pings were taking 10s (or more!) to resolve

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      v1.7.7
Codename:     maroilles
Go version:   go1.11.4
Built:        2019-01-08_10:21:03AM
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
logLevel = "INFO"
InsecureSkipVerify = true

MaxIdleConnsPerHost = 100000
defaultEntryPoints = ["https"]

[file]
  directory = "/etc/traefik-agam/"
  watch = true

[entryPoints]
  [entryPoints.http]
  address = "xxx.xxx.xxx.xxx:80"
  [entryPoints.https]
  address = "xxx.xxx.xxx.xxx:443"
    [entryPoints.https.tls]
  [entryPoints.traefik]
  address = "localhost:65081"

[api]
entryPoint = "traefik"
dashboard = true

[ping]
entryPoint = "http"

[acme]
email = "xxxxxxxx@xxxxxxxx.com"
storage = "/etc/traefik-agam/acme.json"
entryPoint = "https"
onHostRule = true
acmeLogging = true

  [acme.dnsChallenge]
  provider = "cloudflare"

# In another file from /etc/traefik-agam
[backends]
  [backends.foo-bar-backend]
    [backends.foo-bar-backend.servers.server1]
    url = "http://localhost:8071"
    weight = 1
    [backends.foo-bar-s01.servers.server1]
    url = "http://localhost:8072"
    weight = 1
    [backends.foo-bar-s02.servers.server1]
    url = "http://localhost:8073"
    weight = 1
    [backends.foo-bar-s03.servers.server1]
    url = "http://localhost:8074"
    weight = 1
    [backends.foo-bar-s04.servers.server1]
    url = "http://localhost:8075"
    weight = 1
    [backends.foo-bar-s05.servers.server1]
    url = "http://localhost:8076"
    weight = 1
    [backends.foo-bar-s06.servers.server1]
    url = "http://localhost:8077"
    weight = 1
    [backends.foo-bar-s07.servers.server1]
    url = "http://localhost:8078"
    weight = 1
    [backends.foo-bar-s08.servers.server1]
    url = "http://localhost:8079"
    weight = 1
    [backends.foo-bar-s09.servers.server1]
    url = "http://localhost:8068"
    weight = 1
    [backends.foo-bar-adm.servers.server1]
    url = "http://localhost:8067"
    weight = 1

[frontends]
  [frontends.foo-bar-s01]
    backend = "foo-bar-s01"
    entryPoints = [ "https" ]
    [frontends.foo-bar-s01.routes.main]
    rule = "Host: foo.bar; PathPrefix: /s01"
  [frontends.foo-bar-s02]
    backend = "foo-bar-s02"
    entryPoints = [ "https" ]
    [frontends.foo-bar-s02.routes.main]
    rule = "Host: foo.bar; PathPrefix: /s02"
  [frontends.foo-bar-s03]
    backend = "foo-bar-s03"
    entryPoints = [ "https" ]
    [frontends.foo-bar-s03.routes.main]
    rule = "Host: foo.bar; PathPrefix: /s03"
  [frontends.foo-bar-s04]
    backend = "foo-bar-s04"
    entryPoints = [ "https" ]
    [frontends.foo-bar-s04.routes.main]
    rule = "Host: foo.bar; PathPrefix: /s04"
  [frontends.foo-bar-s05]
    backend = "foo-bar-s05"
    entryPoints = [ "https" ]
    [frontends.foo-bar-s05.routes.main]
    rule = "Host: foo.bar; PathPrefix: /s05"
  [frontends.foo-bar-s06]
    backend = "foo-bar-s06"
    entryPoints = [ "https" ]
    [frontends.foo-bar-s06.routes.main]
    rule = "Host: foo.bar; PathPrefix: /s06"
  [frontends.foo-bar-s07]
    backend = "foo-bar-s07"
    entryPoints = [ "https" ]
    [frontends.foo-bar-s07.routes.main]
    rule = "Host: foo.bar; PathPrefix: /s07"
  [frontends.foo-bar-s08]
    backend = "foo-bar-s08"
    entryPoints = [ "https" ]
    [frontends.foo-bar-s08.routes.main]
    rule = "Host: foo.bar; PathPrefix: /s08"
  [frontends.foo-bar-s09]
    backend = "foo-bar-s09"
    entryPoints = [ "https" ]
    [frontends.foo-bar-s09.routes.main]
    rule = "Host: foo.bar; PathPrefix: /s09"
  [frontends.foo-bar-adm]
    backend = "foo-bar-adm"
    entryPoints = [ "https" ]
    [frontends.foo-bar-adm.routes.main]
    rule = "Host: foo.bar; PathPrefix: /adm"

  [frontends.foo-bar-frontend]
  backend = "foo-bar-backend"
  entryPoints = [ "https", "http" ]

    [frontends.foo-bar-frontend.routes.main]
    rule = "Host: foo.bar"
    [frontends.foo-bar-frontend.redirect]
    entryPoint = "https"
```

```
CPU Intel(R) Xeon(R) CPU E3-1275 v5 @ 3.60GHz
System: openSUSE Leap 15.0
```

I would appreciate any tips. I would be happy to capture more logs if needed. For other less demanding services traefik seems to be running fine.

**Comment by MichaelErmer** [comment]
Can you try 1.6.6? 
I'm running a similar setup with 1000's of websockets, everything is fine on 1.6.6, but when trying to upgrade to 1.7 its getting very slow within in a very short time.

**Comment by Rush** [comment]
@MichaelErmer thanks. I'll try. So did the regression occur between 1.6.6 and 1.7.0? Have you tried other 1.7.x versions?

**Comment by MichaelErmer** [comment]
> @MichaelErmer thanks. I'll try. So did the regression occur between 1.6.6 and 1.7.0? Have you tried other 1.7.x versions?

I tried 1.7.5 and 1.7.7

**Comment by Rush** [comment]
I did try 1.6.6 - it was a little better. I wasn't seeing any skyrocketing response times, the page kept loading fine and ping was snappy. Unfortunately the websockets weren't connecting properly. I wasn't able to use my app. Switched back to http-master for the time being. I am open to testing any debug build to get to the bottom of this ... I really like traefik's features.

**Comment by Rush** [comment]
I am using a single backend so it probably won't make a difference. Or did  I misunderstand how stickiness works?

**Comment by Rush** [comment]
@juliens @mmatur does traefik has some kind of a websocket stress test? Maybe writing something like it could expose the underlying issue in an isolated manner.

**Comment by Rush** [comment]
I was able to "crash" my traefik 1.7.7 (systemd watchdog set to 1s) with the following setup.

Please install Node.JS and setup this Node.JS server
https://x.rushbase.net/websocket-test.zip
```
unzip websocket-test.zip
cd websocket-test
npm install
node index.js # listening on port 40000
```

Traefik route config:
```
[backends]
  [backends.test-rushbase-net-backend]
    [backends.test-rushbase-net-backend.servers.server1]
    url = "http://localhost:40000"
    weight = 1

[frontends]
  [frontends.test-somedomain-net-frontend]
  backend = "test-somedomain-net-backend"
  entryPoints = [ "https", "http" ]
  passHostHeader = true

    [frontends.test-somedomain-net-frontend.headers]
    SSLRedirect = true

    [frontends.test-somedomain-net-frontend.routes.main]
    rule = "Host: test.somedomain.net"
```

Please install Artillery stress testing tool
```
npm install -g artillery
```

artillery.yaml (https://rush.sh/LApL+_7Ac+Wn3kwsNAeHNe_TWJQ/artillery.yaml)
```
config:
  target: "wss://test.rushbase.net/s01/"
  phases:
    - duration: 320
      arrivalRate: 100
      rampTo: 2100
  ws:
    rejectUnauthorized: false
scenarios:
  - engine: "ws"
    flow:
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - send: "/room/channel/IDXXXXX"
      - think: 1
      - think: 120
```

run with:
```
artillery run artillery.yaml
```

It seems that <500 connections is sufficient to crash traefik (again, it's a watchdog crash)

Let me know if you have any questions or help in reproducing it!

**Comment by mmatur** [comment]
Hi @Rush,

Thanks for provided us a scenario to reproduce your issue.

Unfortunately we have done some test with @juliens and we haven't been able to reproduce your issue.

Could you please tell us more about the infrastructure on which you have this issue?

**Comment by Rush** [comment]
@mmatur I was able to reproduce the issue on my laptop so architecture doesn't seem to be of relevance. I will try to give you a Dockerfile that reproduces the issue.

**Comment by mmatur** [comment]
Hi @Rush 

Thanks for your feedback. If you are able to give us a Dockerfile that reproduce the issue It could be awesome.

**Comment by Rush** [comment]
Well ... I found the issue and it isn't traefik's fault. Although better error handling would have helped to identify the problem. In short, openSUSE has max open files was set by default to 1024. It was working fine with http-master with its multi-worker design with which the total connection limit was 8*1024 (where 8 is the number of workers). Traefik on the other hand has one process and 2100 concurrent connections was quickly exhausting the limit. 

If you would like to reproduce such connection limit locally, here's a docker-compose setup https://rush.sh/kx73DK7YALGj_QEYr+V0idr_APs/websocket-issue.zip (the behavior in docker is not exactly the same as outside it but it's close enough)

To fix my setup, I have added this additional setting into traefik's service file:
```
[Service]
...
LimitNOFILE=256380
```

I propose to create another issue for reporting a lack of available file descriptors as an error.
